op-node/rollup/derive/batches.go

```diff
@@ -25,15 +26,95 @@ const (
+	jst := batch.Batch.Justification
+	if jst == nil {
+		// Espresso blocks must have a justification
+		return BatchDrop
+	}
+	startHeaders :=
+		[]espresso.Header{jst.PrevBatchLastBlock, jst.FirstBlock}
+	headerHeights :=
+		[]uint64{jst.FirstBlockNumber - 1, jst.FirstBlockNumber}
+	err := hotshot.verifyHeaders(startHeaders, headerHeights)
+
+	// In the case that the headers aren't available yet (perhaps the validator's L1 client is behind), return BatchFuture so that we can try again later
+	// If the headers are available but invalid, drop the batch
+	if err != nil {
+		return BatchFuture
+		// TODO drop the batch if the header is invalid instead of unavailable
+	}
+
+	// First, check for cases where it is valid to have any empty batch.
+	windowStart := l2SafeHead.Time + cfg.BlockTime
+	windowEnd := windowStart + cfg.BlockTime
+	payload := jst.Payload
+	prevL1Origin := l1Blocks[0]
+
+	// If Espresso did not produce any blocks in this window, an empty batch is valid.
+	// In this case, the L1 origin must be the same as the previous block.
+	if payload == nil && jst.FirstBlock.Timestamp >= windowEnd {
+		if uint64(batch.Batch.EpochNum) != prevL1Origin.Number {
+			return BatchDrop
+		}
+		return BatchAccept
+	}
+
+	// An empty batch can also be valid if the L1 origin is too far behind (either due to lag, or because HotShot skipped a block)
+	// In these cases, the L1 origin must increase by one
+	// TODO check lag case
+	skippedL1Block := jst.FirstBlock.L1Block.Number > jst.PrevBatchLastBlock.L1Block.Number+1
+	if payload == nil && skippedL1Block {
+		if uint64(batch.Batch.EpochNum) != prevL1Origin.Number+1 {
+			return BatchDrop
+		}
+		return BatchAccept
+	}
+
+	// At this point, there is no valid reason to have an empty payload
+	if payload == nil {
+		return BatchDrop
+	}
+
+	// Now validate the transactions in the batch
+	numBlocks := len(payload.NmtProofs)
+
+	// Validate the headers representing the end of the batch window
+	endHeaders :=
+		[]espresso.Header{payload.LastBlock, payload.NextBatchFirstBlock}
+	headerHeights =
+		[]uint64{jst.FirstBlockNumber + uint64(numBlocks), jst.FirstBlockNumber + uint64(numBlocks) + 1}
+	err = hotshot.verifyHeaders(endHeaders, headerHeights)
+
+	// In the case that the headers aren't available yet (perhaps the validator's L1 client is behind), return BatchFuture so that we can try again later
+	// If the headers are available but invalid, drop the batch
+	if err != nil {
+		// TODO drop the batch if there is a true validation error
+		return BatchFuture
+	}
+
+	// Check that the range of HotShot blocks fall within the window
+	validRange :=
+		jst.PrevBatchLastBlock.Timestamp < windowStart &&
+			jst.FirstBlock.Timestamp >= windowStart &&
+			payload.LastBlock.Timestamp < windowEnd &&
+			payload.NextBatchFirstBlock.Timestamp >= windowEnd
+
+	if !validRange {
+		return BatchDrop
+	}
+
+	// TODO validate NMT proofs
```

Comment:
Yeah, let's add mock validation functions with the proper interface that always succeed in the `espresso` package. That way the logic here will look as it should and the place where we're cheating will be self-contained.